File: Sources/redline/ClaudeCredentialSource.swift
// Reads the credential Claude Code already stored, in increasing order of intrusion: a plain
// file, then the Apple-signed `security` tool, then the Keychain API last because it is the
// one path that shows a consent prompt. Every read reports why it failed, because "signed
// out" and "not allowed yet" need opposite responses.
import Foundation
import RedlineCore
import RedlineUI
import Security

enum ClaudeCredentialSource {
    static let keychainService = "Claude Code-credentials"

    /// Tried in order of intrusion. The file costs nothing and prompts for nothing. The
    /// `security` binary sits in the item's apple-tool: partition, so on most installs it
    /// reads without any consent UI; this is how every peer tool avoids the recurring
    /// password prompt, and Claude Code rewriting the item cannot revoke it. The direct API
    /// call is last: it is the only rung that prompts, and the prompt has to be re-granted
    /// every time Claude Code rewrites the item's ACL.
    /// Blocks on the Keychain consent prompt, so never call from the main thread.
    static func load(home: URL? = nil,
                     allowSecurityCLI: Bool = true) -> CredentialOutcome {
        let fileOutcome = fromFile(home: home)
        if case .found = fileOutcome { return fileOutcome }

        var cliOutcome: CredentialOutcome = .notFound
        if allowSecurityCLI {
            cliOutcome = fromSecurityCLI()
            if case .found = cliOutcome { return cliOutcome }
        }

        // The prompting path, reached only when the quiet ones came back empty-handed
        let keychainOutcome = fromKeychain()
        if case .found = keychainOutcome { return keychainOutcome }

        // Prefer the more specific answer: a present-but-locked item is not an absent one
        if keychainOutcome == .notFound, cliOutcome == .notFound, fileOutcome == .notFound {
            return .notFound
        }
        return keychainOutcome == .notFound ? cliOutcome : keychainOutcome
    }

    // MARK: - File

    static func fromFile(home: URL? = nil) -> CredentialOutcome {
        let root = home ?? RedlineHome.url
        let path = root.appendingPathComponent(".claude/.credentials.json")
        guard let data = try? Data(contentsOf: path) else { return .notFound }
        return parse(data)
    }

    // MARK: - Keychain API

    static func fromKeychain() -> CredentialOutcome {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: keychainService,
            kSecReturnData: true,
            kSecMatchLimit: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        switch status {
        case errSecSuccess:
            guard let data = item as? Data else { return .unreadable }
            return parse(data)
        case errSecItemNotFound:
            return .notFound
        default:
            // errSecInteractionNotAllowed, errSecAuthFailed, errSecUserCanceled and friends all
            // mean "exists, not handed over". Discarding this status is what latched the old
            // probe at nil until the user clicked Reconnect.
            return .accessDenied
        }
    }

    /// Attribute-only read: no kSecReturnData, so the secret is never decrypted and the ACL is
    /// never consulted. Free way to notice that Claude Code rotated the item.
    static func keychainModifiedAt() -> Date? {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: keychainService,
            kSecReturnAttributes: true,
            kSecMatchLimit: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let attrs = item as? [CFString: Any] else { return nil }
        return attrs[kSecAttrModificationDate] as? Date
    }

    // MARK: - security(1)

    static func fromSecurityCLI(account: String? = nil) -> CredentialOutcome {
        var args = ["find-generic-password", "-s", keychainService]
        if let account { args += ["-a", account] }
        args.append("-w")

        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: "/usr/bin/security")
        proc.arguments = args
        let out = Pipe()
        let err = Pipe()
        proc.standardOutput = out
        proc.standardError = err
        let launched: Void? = Diag.log.attempt("keychain.security_launch_failed") {
            try proc.run()
        }
        guard launched != nil else { return .accessDenied }

        // The consent dialog blocks until answered. A short kill here reads a waiting user as
        // a denial and reports "disconnected" while the prompt is still on screen.
        let deadline = Date().addingTimeInterval(90)
        while proc.isRunning, Date() < deadline { usleep(50_000) }
        if proc.isRunning {
            proc.terminate()
            return .accessDenied
        }

        let stdout = out.fileHandleForReading.readDataToEndOfFile()
        let stderr = String(data: err.fileHandleForReading.readDataToEndOfFile(),
                            encoding: .utf8)?.lowercased() ?? ""
        guard proc.terminationStatus == 0 else {
            // 44 is security(1)'s "item not found"; anything else means it is there but shut
            let missing = proc.terminationStatus == 44
                || stderr.contains("could not be found")
            return missing ? .notFound : .accessDenied
        }
        let text = SecurityCLIOutput.decode(String(data: stdout, encoding: .utf8) ?? "")
        guard let data = text.data(using: .utf8) else { return .unreadable }
        return parse(data)
    }

    // MARK: -

    private static func parse(_ data: Data) -> CredentialOutcome {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            Diag.log.error("keychain.credential_unparsed",
                           "Keychain payload was not the expected JSON",
                           ["bytes": "\(data.count)"])
            return .unreadable
        }
        guard let credential = CredentialScan.credential(in: json) else { return .unreadable }
        return .found(credential)
    }
}

File: windows/src/Redline.App/Services/ClaudeCredentialSource.cs
// Reads the credential Claude Code already stored, least intrusive first: the plain file Claude
// Code uses on Windows, then Credential Manager. Read only; every miss says why it missed.
using System.IO;
using System.Text;
using System.Text.Json.Nodes;
using Redline.Core;

namespace Redline.App.Services;

public static class ClaudeCredentialSource
{
    /// The service name Claude Code uses for its Keychain item, and so for any keytar-style store.
    public const string CredentialTarget = "Claude Code-credentials";

    public static string FilePath(string? home = null) =>
        RedlineHome.Join(home ?? RedlineHome.Url, ".claude/.credentials.json");

    /// The file first, because on Windows that is where Claude Code keeps it. Credential Manager
    /// never prompts, so the macOS `security` rung has nothing to replace and is dropped.
    public static CredentialOutcome Load(string? home = null, ICredentialStore? store = null)
    {
        var fileOutcome = FromFile(home);
        if (fileOutcome is CredentialOutcome.Found) return fileOutcome;

        var storeOutcome = FromCredentialManager(store ?? WindowsCredentialStore.Shared);
        if (storeOutcome is CredentialOutcome.Found) return storeOutcome;

        // Prefer the more specific answer: a present-but-refused item is not an absent one
        if (storeOutcome is CredentialOutcome.NotFound && fileOutcome is CredentialOutcome.NotFound)
            return new CredentialOutcome.NotFound();
        return storeOutcome is CredentialOutcome.NotFound ? fileOutcome : storeOutcome;
    }

    public static CredentialOutcome FromFile(string? home = null)
    {
        var path = FilePath(home);
        if (!File.Exists(path)) return new CredentialOutcome.NotFound();
        byte[] data;
        try { data = File.ReadAllBytes(path); }
        // Present but locked or unreadable is transient, never "signed out"
        catch (UnauthorizedAccessException) { return new CredentialOutcome.AccessDenied(); }
        catch (IOException) { return new CredentialOutcome.AccessDenied(); }
        return Parse(data);
    }

    /// The exact target first, then any `Claude Code-credentials*` item a keytar build suffixed.
    public static CredentialOutcome FromCredentialManager(ICredentialStore store)
    {
        CredentialRead read;
        try { read = store.Read(CredentialTarget); }
        catch { return new CredentialOutcome.AccessDenied(); }
        if (read.Status == CredentialReadStatus.Found && read.Credential is { } c)
            return Parse(Encoding.UTF8.GetBytes(WindowsCredentialStore.DecodeBlob(c.Blob)));
        if (read.Status == CredentialReadStatus.AccessDenied) return new CredentialOutcome.AccessDenied();

        IReadOnlyList<StoredCredential> matches;
        try { matches = store.Enumerate(CredentialTarget + "*"); }
        catch { return new CredentialOutcome.NotFound(); }
        CredentialOutcome best = new CredentialOutcome.NotFound();
        foreach (var m in matches.OrderBy(m => m.Target, StringComparer.Ordinal))
        {
            var outcome = Parse(Encoding.UTF8.GetBytes(WindowsCredentialStore.DecodeBlob(m.Blob)));
            if (outcome is CredentialOutcome.Found) return outcome;
            best = outcome;
        }
        return best;
    }

    /// When the credential last changed, read without parsing it: a moved stamp means Claude
    /// Code rotated the token and the cached copy is worthless.
    public static DateTimeOffset? ModifiedAt(string? home = null, ICredentialStore? store = null)
    {
        try
        {
            var path = FilePath(home);
            if (File.Exists(path)) return new DateTimeOffset(File.GetLastWriteTimeUtc(path), TimeSpan.Zero);
            var read = (store ?? WindowsCredentialStore.Shared).Read(CredentialTarget);
            return read.Credential?.LastWritten;
        }
        catch { return null; }
    }

    internal static CredentialOutcome Parse(byte[] data)
    {
        JsonObject? json = null;
        try { json = JsonNode.Parse(data) as JsonObject; } catch { }
        if (json is null)
        {
            Diag.Log.Error("credential.unparsed", "stored credential was not the expected JSON",
                new() { ["bytes"] = data.Length.ToString() });
            return new CredentialOutcome.Unreadable();
        }
        return CredentialScan.Credential(json) is { } credential
            ? new CredentialOutcome.Found(credential)
            : new CredentialOutcome.Unreadable();
    }
}
